zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -0,0 +1,521 @@
+            OpcodeId::AND => ExecutionState::BITWISE,
+            OpcodeId::XOR => ExecutionState::BITWISE,
+            OpcodeId::OR => ExecutionState::BITWISE,
+            OpcodeId::POP => ExecutionState::POP,
+            OpcodeId::PUSH32 => ExecutionState::PUSH,
+            OpcodeId::BYTE => ExecutionState::BYTE,
+            OpcodeId::MLOAD => ExecutionState::MEMORY,
+            OpcodeId::MSTORE => ExecutionState::MEMORY,
+            OpcodeId::MSTORE8 => ExecutionState::MEMORY,
+            OpcodeId::JUMPDEST => ExecutionState::JUMPDEST,
+            OpcodeId::PC => ExecutionState::PC,
+            _ => unimplemented!("unimplemented opcode {:?}", step.op),
+        }
+    }
+}
+
+impl From<&bus_mapping::bytecode::Bytecode> for Bytecode {
+    fn from(b: &bus_mapping::bytecode::Bytecode) -> Self {
+        Bytecode::new(b.to_bytes())
+    }
+}
+
+fn step_convert(
+    prev: Option<&bus_mapping::circuit_input_builder::ExecStep>,
+    step: &bus_mapping::circuit_input_builder::ExecStep,
+    ops_len: (usize, usize, usize),
+) -> ExecStep {
+    let (stack_ops_len, memory_ops_len, _storage_ops_len) = ops_len;
+    let result = ExecStep {
+        rw_indices: step
+            .bus_mapping_instance
+            .iter()
+            .map(|x| {
+                let index = x.as_usize() - 1;
+                match x.target() {
+                    bus_mapping::operation::Target::Stack => index,
+                    bus_mapping::operation::Target::Memory => {
+                        index + stack_ops_len
+                    }
+                    bus_mapping::operation::Target::Storage => {
+                        index + stack_ops_len + memory_ops_len
+                    }
+                }
+            })
+            .collect(),
+        execution_state: ExecutionState::from(step),
+        rw_counter: usize::from(step.gc),
+        program_counter: usize::from(step.pc) as u64,
+        stack_pointer: 1024 - step.stack_size,
+        gas_left: step.gas_left.0,
+        gas_cost: step.gas_cost.as_u64(),
+        opcode: Some(step.op),
+        // As in https://github.com/ethereum/go-ethereum/blob/bc6bf1e1937829b5d5b0d431a9333d47c3e08082/core/vm/interpreter.go#L224-L233,
+        // geth increases memory size before making trace,
+        // so the memory size in ExecStep is not what we expect to see.
+        // We have to use the memory size of previous step as correct memory
+        // size before each step
+        memory_size: match prev {
+            None => 0,
+            Some(prev_step) => (prev_step.memory_size as u64) / 32, /* memory size in word */
+        },
+        ..Default::default()
+    };
+    result
+}
+
+fn tx_convert(
+    randomness: Fp,
+    bytecode: &Bytecode,
+    tx: &bus_mapping::circuit_input_builder::Transaction,
+    ops_len: (usize, usize, usize),
+) -> Transaction<Fp> {
+    let mut result: Transaction<Fp> = Transaction::<Fp> {
+        calls: vec![Call {
```

Comment:
I see that you're defining the implicit call generated from the transaction here.  I just wanted to mention that this initial call can also be found in `tx.calls[0]` See https://github.com/appliedzkp/zkevm-circuits/blob/b92b2953df5709d8e2c8ce7090960b13394efbc0/bus-mapping/src/circuit_input_builder.rs#L348
The `tx.calls` is private, and there's pub method to access them, but there should be one!  I just wanted to mention this so that in the future the circuits can use the calls values generated by the CircuitInputBuilder.